lib/kafka_ex/gen_consumer.ex

```diff
@@ -0,0 +1,612 @@
+      {:async_commit, %{state | messages: state.messages ++ message_set}}
+    end
+
+    def handle_call(:messages, _from, messages_so_far)
+      {:reply, state.messages, %{state | calls: state.calls + 1}}
+    end
+  end
+
+  {:ok, pid} = GenConsumer.start_link(MyConsumer, "consumer_group", "topic", 0)
+  GenConsumer.call(pid, :messages)
+  ```
+
+  **NOTE** If you do not implement a `c:handle_call/3` callback, any calls to
+  `GenConsumer.call/3` that go to your consumer will cause a `MatchError`.
+
+  ## Testing
+
+  A `KafkaEx.GenConsumer` can be unit-tested without a running Kafka broker by sending
+  messages directly to its `c:handle_message_set/2` function. The following
+  recipe can be used as a starting point when testing a `KafkaEx.GenConsumer`:
+
+  ```
+  defmodule ExampleGenConsumerTest do
+    use ExUnit.Case, async: true
+
+    alias KafkaEx.Protocol.Fetch.Message
+
+    @topic "topic"
+    @partition 0
+
+    setup do
+      {:ok, state} = ExampleGenConsumer.init(@topic, @partition)
+      {:ok, %{state: state}}
+    end
+
+    test "it acks a message", %{state: state} do
+      message_set = [%Message{offset: 0, value: "hello"}]
+      {response, _new_state} =
+        ExampleGenConsumer.handle_message_set(message_set, state)
+      assert response == :async_commit
+    end
+  end
+  ```
+  """
+
+  use GenServer
+
+  alias KafkaEx.Protocol.OffsetCommit.Request, as: OffsetCommitRequest
+  alias KafkaEx.Protocol.OffsetCommit.Response, as: OffsetCommitResponse
+  alias KafkaEx.Protocol.OffsetFetch.Request, as: OffsetFetchRequest
+  alias KafkaEx.Protocol.OffsetFetch.Response, as: OffsetFetchResponse
+  alias KafkaEx.Protocol.Offset.Response, as: OffsetResponse
+  alias KafkaEx.Protocol.Fetch.Response, as: FetchResponse
+  alias KafkaEx.Protocol.Fetch.Message
+
+  require Logger
+
+  @typedoc """
+  Option values used when starting a `KafkaEx.GenConsumer`.
+  """
+  @type option :: {:commit_interval, non_neg_integer}
+                | {:commit_threshold, non_neg_integer}
+
+  @typedoc """
+  Options used when starting a `KafkaEx.GenConsumer`.
+  """
+  @type options :: [option | GenServer.option]
+
+  @doc """
+  Invoked when the server is started. `start_link/5` will block until it
+  returns.
+
+  `topic` and `partition` are the arguments passed to `start_link/5`. They
+  identify the Kafka partition that the `KafkaEx.GenConsumer` will consume from.
+
+  Returning `{:ok, state}` will cause `start_link/5` to return `{:ok, pid}` and
+  the process to start consuming from its assigned partition. `state` becomes
+  the consumer's state.
+
+  Any other return value will cause the `start_link/5` to return `{:error,
+  error}` and the process to exit.
+  """
+  @callback init(topic :: binary, partition :: non_neg_integer) ::
+    {:ok, state :: term}
+
+  @doc """
+  Invoked for each message set consumed from a Kafka topic partition.
+
+  `message_set` is a message set fetched from a Kafka broker and `state` is the
```

Comment:
maybe say "list" instead of "set", or even just "Enumerable", otherwise it sort of sounds like it's a MapSet or something.